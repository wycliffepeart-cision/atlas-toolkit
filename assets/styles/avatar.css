@layer atlas {
	atlas-avatar {
		--atlas-avatar-vertex-clip-path: circle(50%);
		--atlas-avatar-default-width: 2em;
		--atlas-avatar-full-mode-min-width: 2em;
	}

	atlas-avatar::part(image),
	atlas-avatar::part(initials),
	atlas-avatar::part(fallback) {
		display: none;
		aspect-ratio: 1 / 1;
		width: var(--atlas-avatar-default-width);
		clip-path: var(--atlas-avatar-vertex-clip-path);
	}

	atlas-avatar[visible="image"]::part(image),
	atlas-avatar[visible="initials"]::part(initials),
	atlas-avatar[visible="fallback"]::part(fallback) {
		display: flex;
		align-items: center;
		justify-content: center;
	}

	atlas-avatar::part(image) {
		position: relative;
		object-fit: cover;
	}

	atlas-avatar[full-mode][visible="initials"],
	atlas-avatar[full-mode][visible="fallback"] {
		min-width: var(--atlas-avatar-full-mode-min-width);
		height: auto;
	}

	atlas-avatar[full-mode][visible="image"]::part(image),
	atlas-avatar[full-mode][visible="initials"]::part(initials),
	atlas-avatar[full-mode][visible="fallback"]::part(fallback) {
		aspect-ratio: unset;
		--atlas-avatar-vertex-clip-path: inset(0% 0% 0% 0% round 0.3em);
	}

	atlas-avatar[full-mode][visible="initials"]::part(initials),
	atlas-avatar[full-mode][visible="fallback"]::part(fallback) {
		width: 100%;
		height: 100%;
	}
}
