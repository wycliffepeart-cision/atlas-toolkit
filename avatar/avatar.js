import { _ as __decorate, a as __classPrivateFieldGet } from '../tslib.es6-806de115.js';
import { Template } from '../decorators/template.js';

var AvatarTemplate = "<template>\n  <img id=\"image\" part=\"image\">\n  <div id=\"initials\" part=\"initials\"></div>\n  <slot id=\"fallback\" part=\"fallback\"></slot>\n</template>\n";

var _Avatar_instances, _Avatar_content, _Avatar_defineContent;
const AVATAR_ATTRIBUTES = {
    IMAGE_SOURCE: 'src',
    INITIALS: 'initials',
    VISIBLE: 'visible',
};
const DISPLAY = {
    src: 'image',
    initials: 'initials',
    fallback: 'fallback',
};
const priorityOrder = ['src', 'initials', 'fallback'];
let Avatar = class Avatar extends HTMLElement {
    constructor() {
        super(...arguments);
        _Avatar_instances.add(this);
        _Avatar_content.set(this, {
            fallback: true,
        });
    }
    connectedCallback() {
        const currentVisibleValue = this.getAttribute(AVATAR_ATTRIBUTES.VISIBLE);
        this.setAttribute(AVATAR_ATTRIBUTES.VISIBLE, currentVisibleValue !== null && currentVisibleValue !== void 0 ? currentVisibleValue : DISPLAY.fallback);
    }
    static get observedAttributes() {
        return [AVATAR_ATTRIBUTES.IMAGE_SOURCE, AVATAR_ATTRIBUTES.INITIALS];
    }
    get visible() {
        return this.getAttribute(AVATAR_ATTRIBUTES.VISIBLE);
    }
    attributeChangedCallback(name, _, newValue) {
        __classPrivateFieldGet(this, _Avatar_instances, "m", _Avatar_defineContent).call(this, name, newValue);
        const visibleAttr = priorityOrder.find((attr) => __classPrivateFieldGet(this, _Avatar_content, "f")[attr]);
        this.setAttribute(AVATAR_ATTRIBUTES.VISIBLE, DISPLAY[visibleAttr]);
    }
};
_Avatar_content = new WeakMap();
_Avatar_instances = new WeakSet();
_Avatar_defineContent = function _Avatar_defineContent(name, value) {
    const imageElement = this.shadowRoot.querySelector('#image');
    const initialsElement = this.shadowRoot.querySelector('#initials');
    if (name === AVATAR_ATTRIBUTES.INITIALS) {
        initialsElement.innerHTML = value;
    }
    if (name === AVATAR_ATTRIBUTES.IMAGE_SOURCE) {
        imageElement.setAttribute(AVATAR_ATTRIBUTES.IMAGE_SOURCE, value);
    }
    __classPrivateFieldGet(this, _Avatar_content, "f")[name] = value;
};
Avatar = __decorate([
    Template(AvatarTemplate)
], Avatar);

export { AVATAR_ATTRIBUTES, Avatar, DISPLAY };
