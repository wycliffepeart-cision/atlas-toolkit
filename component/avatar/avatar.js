import { _ as __decorate } from '../../tslib.es6-2e0713a3.js';
import { Template } from '../../decorators/template.js';

var AvatarTemplate = "<template>\n  <img id=\"image\" part=\"image\">\n  <div id=\"initials\" part=\"initials\"></div>\n  <slot id=\"fallback\" part=\"fallback\"></slot>\n</template>\n";

const AVATAR_ATTRIBUTES = {
    IMAGE_SOURCE: 'src',
    INITIALS: 'initials',
    VISIBLE: 'visible',
};
const DISPLAY = {
    src: 'image',
    initials: 'initials',
    fallback: 'fallback',
};
const priorityOrder = ['src', 'initials', 'fallback'];
let Avatar = class Avatar extends HTMLElement {
    #content = {
        fallback: true,
    };
    connectedCallback() {
        const currentVisibleValue = this.getAttribute(AVATAR_ATTRIBUTES.VISIBLE);
        this.setAttribute(AVATAR_ATTRIBUTES.VISIBLE, currentVisibleValue ?? DISPLAY.fallback);
        const imageElement = this.shadowRoot.querySelector('#image');
        imageElement.addEventListener('error', this.#handleImageError);
    }
    disconnectedCallback() {
        const imageElement = this.shadowRoot.querySelector('#image');
        imageElement.removeEventListener('error', this.#handleImageError);
    }
    static get observedAttributes() {
        return [AVATAR_ATTRIBUTES.IMAGE_SOURCE, AVATAR_ATTRIBUTES.INITIALS];
    }
    get visible() {
        return this.getAttribute(AVATAR_ATTRIBUTES.VISIBLE);
    }
    #handleImageError = () => {
        this.dispatchEvent(new CustomEvent('error', { bubbles: true, composed: true }));
    };
    #defineContent(name, value) {
        const imageElement = this.shadowRoot.querySelector('#image');
        const initialsElement = this.shadowRoot.querySelector('#initials');
        if (name === AVATAR_ATTRIBUTES.INITIALS) {
            initialsElement.innerHTML = value;
        }
        if (name === AVATAR_ATTRIBUTES.IMAGE_SOURCE) {
            imageElement.setAttribute(AVATAR_ATTRIBUTES.IMAGE_SOURCE, value);
        }
        this.#content[name] = value;
    }
    attributeChangedCallback(name, _, newValue) {
        this.#defineContent(name, newValue);
        const visibleAttr = priorityOrder.find((attr) => this.#content[attr]);
        this.setAttribute(AVATAR_ATTRIBUTES.VISIBLE, DISPLAY[visibleAttr]);
    }
};
Avatar = __decorate([
    Template(AvatarTemplate)
], Avatar);
if (!customElements.get('atlas-avatar'))
    customElements.define('atlas-avatar', Avatar);

export { AVATAR_ATTRIBUTES, Avatar, DISPLAY };
