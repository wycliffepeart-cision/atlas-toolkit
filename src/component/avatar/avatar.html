<template>
  <img id="image" part="image" />
  <div id="initials" part="initials"></div>
  <slot id="fallback" part="fallback"></slot>
</template>
