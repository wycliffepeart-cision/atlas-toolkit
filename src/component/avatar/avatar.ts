import { Template } from '../../decorators/template';
import AvatarTemplate from './avatar.html';

export const AVATAR_ATTRIBUTES = {
  IMAGE_SOURCE: 'src',
  INITIALS: 'initials',
  VISIBLE: 'visible'
}

export const DISPLAY = {
  src: 'image',
  initials: 'initials',
  fallback: 'fallback'
}

const priorityOrder = ['src', 'initials', 'fallback']

@Template(AvatarTemplate)
export class Avatar extends HTMLElement {

  #content = {
    fallback: true
  }

  connectedCallback() {
    const currentVisibleValue = this.getAttribute(AVATAR_ATTRIBUTES.VISIBLE)
    this.setAttribute(AVATAR_ATTRIBUTES.VISIBLE, currentVisibleValue ?? DISPLAY.fallback)

    const imageElement = this.shadowRoot.querySelector('#image')
    imageElement.addEventListener('error', this.#handleImageError)
  }

  disconnectedCallback() {
    const imageElement = this.shadowRoot.querySelector('#image')
    imageElement.removeEventListener('error', this.#handleImageError)
  }

  static get observedAttributes() {
    return [AVATAR_ATTRIBUTES.IMAGE_SOURCE, AVATAR_ATTRIBUTES.INITIALS]
  }

  get visible() {
    return this.getAttribute(AVATAR_ATTRIBUTES.VISIBLE)
  }

  #handleImageError = () => {
    this.dispatchEvent(new CustomEvent('error', { bubbles: true, composed: true }))
  }

  #defineContent(name, value) {
    const imageElement = this.shadowRoot.querySelector('#image')
    const initialsElement = this.shadowRoot.querySelector('#initials')

    if (name === AVATAR_ATTRIBUTES.INITIALS) {
      initialsElement.innerHTML = value
    }

    if (name === AVATAR_ATTRIBUTES.IMAGE_SOURCE) {
      imageElement.setAttribute(AVATAR_ATTRIBUTES.IMAGE_SOURCE, value)
    }

    this.#content[name] = value
  }

  attributeChangedCallback(name, _, newValue) {
    this.#defineContent(name, newValue)

    const visibleAttr = priorityOrder.find(attr => this.#content[attr])

    this.setAttribute(AVATAR_ATTRIBUTES.VISIBLE, DISPLAY[visibleAttr])
  }
}

if (!customElements.get('atlas-avatar')) customElements.define('atlas-avatar', Avatar);
