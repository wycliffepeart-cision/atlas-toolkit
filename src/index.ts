export * from './component/accordion/accordion';
export * from './component/alert/alert';
export * from './component/application-frame/application-frame';
export * from './component/button/button';
export * from './component/icon/icon';
export * from './component/button-set/button-set';
export * from './component/chip/chip';
export * from './component/chip-group/chip-group';
export * from './component/dialog/dialog';
export * from './component/feed/feed';
export * from './component/input-checkbox/input-checkbox';
export * from './component/input-chip-group/input-chip-group';
export * from './component/input-dropdown/input-dropdown';
export * from './component/input-search/input-search';
export * from './component/input-text/input-text';
export * from './component/input-textarea/input-textarea';
export * from './component/logo/logo';
export * from './component/main-nav/main-nav';
export * from './component/main-nav-item/main-nav-item';
export * from './component/number-display/number-display';
export * from './component/profile-control/profile-control';
export * from './component/snackbar/snackbar';
export * from './component/stepper/stepper';
export * from './component/stepper-step/stepper-step';
export * from './component/tile-board/tile-board';
export * from './component/tile-board-tile/tile-board-tile';
export * from './component/workflow/workflow';
export * from './component/workflow-step/workflow-step';
