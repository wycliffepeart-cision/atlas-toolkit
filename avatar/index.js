import { Avatar } from './avatar.js';
import '../tslib.es6-806de115.js';
import '../decorators/template.js';

if (!customElements.get('atlas-avatar'))
    customElements.define('atlas-avatar', Avatar);
