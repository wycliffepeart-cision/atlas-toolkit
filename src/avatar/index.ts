import { Avatar } from './avatar';

if (!customElements.get('atlas-avatar')) customElements.define('atlas-avatar', Avatar);
