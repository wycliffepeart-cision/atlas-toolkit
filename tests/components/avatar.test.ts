import '../../src/avatar';

const ELEMENT_TAG = 'atlas-avatar';
const VISIBLE_ATTRIBUTE_NAME = 'visible';
const getShadowDomContent = (element) => element.shadowRoot.innerHTML;
const shadowContentSnapshot = `
<template>
    <img id="image" part="image">
    <div id="initials" part="initials"></div>
    <slot id="fallback" part="fallback" visible="true">
    </slot>
</template>
`;

afterEach(() => {
  document.body.innerHTML = '';
});

describe('Avatar', () => {
  it('is correctly appended in the DOM', async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    (expect(element) as any).toBeInTheDocument();
  });

  it('is capable of being selected', async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    const selectedElement = document.querySelector(ELEMENT_TAG);
    (expect(selectedElement) as any).toBeInTheDocument();
  });

  it('has the correct content when created', async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    (expect(element) as any).toContainHTML('<atlas-avatar visible="fallback"></atlas-avatar>');
    (expect(element) as any).toBeEmptyDOMElement();
    (expect(getShadowDomContent(element)) as any).toMatchSnapshot(shadowContentSnapshot);
  });

  it("sets the fallback element as visible when the 'src' nor the 'initials' attributes are set in the HTML tag", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const htmlStringContainer = document.createElement('div');
    document.body.appendChild(htmlStringContainer);

    htmlStringContainer.innerHTML = `<atlas-avatar></atlas-avatar>`;

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('fallback');
  });

  it("sets the fallback element as visible when the 'src' nor the 'initials' attributes are set programmatically", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);

    document.body.appendChild(element);

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('fallback');
  });

  it("sets the user image as visible when only the 'src' attribute is set in the HTML tag", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const htmlStringContainer = document.createElement('div');
    document.body.appendChild(htmlStringContainer);

    htmlStringContainer.innerHTML = `<atlas-avatar
			src="https://randomuser.me/api/portraits/men/11.jpg">
		</atlas-avatar>`;

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('image');
  });

  it("sets the user image as visible when only the 'src' attribute is set programmatically", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    element.setAttribute('src', 'https://randomuser.me/api/portraits/men/11.jpg');

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('image');
  });

  it("sets the user initials as visible when only the 'initials' attribute is set in the HTML tag", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const htmlStringContainer = document.createElement('div');
    document.body.appendChild(htmlStringContainer);

    htmlStringContainer.innerHTML = `<atlas-avatar initials="CO"></atlas-avatar>`;

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('initials');
  });

  it("sets the user initials as visible when only the 'initials' attribute is set programmatically", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    element.setAttribute('initials', 'CO');

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('initials');
  });

  it("sets only the user image as visible when the 'src' and 'initials' attributes are set in the HTML tag", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const htmlStringContainer = document.createElement('div');
    document.body.appendChild(htmlStringContainer);

    htmlStringContainer.innerHTML = `<atlas-avatar initials="CO"
			src="https://randomuser.me/api/portraits/men/11.jpg">
		</atlas-avatar>`;

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('image');
  });

  it("sets only the user image as visible when the 'src' and 'initials' attributes are set programmatically", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    element.setAttribute('initials', 'CO');
    element.setAttribute('src', 'https://randomuser.me/api/portraits/men/11.jpg');

    const avatar = document.querySelector(ELEMENT_TAG);
    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('image');
  });

  it("sets only the fallback element as visible when both the 'src' and 'initials' attributes have been set and then removed programatically", async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);
    element.setAttribute('initials', 'CO');
    element.setAttribute('src', 'https://randomuser.me/api/portraits/men/11.jpg');
    element.removeAttribute('initials');
    element.removeAttribute('src');

    const avatar = document.querySelector(ELEMENT_TAG);

    (expect(avatar.getAttribute(VISIBLE_ATTRIBUTE_NAME)) as any).toBe('fallback');
  });

  it('returns visible values as shortcut to attribute before set', async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element: any = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    (expect(element.visible) as any).toEqual('fallback');
  });

  it('returns visible values as shortcut to attribute after set', async () => {
    await customElements.whenDefined(ELEMENT_TAG);
    const element: any = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    element.setAttribute('initials', 'XY');

    (expect(element.visible) as any).toEqual('initials');
  });
});
