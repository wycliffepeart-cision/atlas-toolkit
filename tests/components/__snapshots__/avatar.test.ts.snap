// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar has the correct content when created: 
<template>
    <img id="image" part="image">
    <div id="initials" part="initials"></div>
    <slot id="fallback" part="fallback" visible="true">
    </slot>
</template>
 1`] = `
<img id="image"
     part="image"
>
<div id="initials"
     part="initials"
>
</div>
<slot id="fallback"
      part="fallback"
>
</slot>
`;
